zellij-server/src/decorating_pipe.rs

```diff
@@ -0,0 +1,94 @@
+use std::{
+    collections::VecDeque,
+    io::{Read, Seek, Write},
+};
+
+use log::info;
+use serde::{Deserialize, Serialize};
+use wasmer_wasi::{WasiFile, WasiFsError};
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct DecoratingPipe {
+    buffer: VecDeque<u8>,
+    plugin_name: String,
+}
+
+impl DecoratingPipe {
+    pub fn new(plugin_name: &str) -> DecoratingPipe {
+        info!("Creating decorating pipe!");
+        dbg!("Creating the decorating pipe :)");
+        DecoratingPipe {
+            buffer: VecDeque::new(),
+            plugin_name: String::from(plugin_name),
+        }
+    }
+}
+
+impl Read for DecoratingPipe {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        let amt = std::cmp::min(buf.len(), self.buffer.len());
+        for (i, byte) in self.buffer.drain(..amt).enumerate() {
+            buf[i] = byte;
+        }
+        Ok(amt)
+    }
+}
+
+// TODO: do this better. We're not sure about byte boundaries and endl stuff but, we do expect
+// to get the valid thing eventually
+impl Write for DecoratingPipe {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        self.buffer.extend(buf);
+        let current_chunk = std::str::from_utf8(buf).unwrap();
+        for c in current_chunk.chars() {
+            if c == '\n' {
+                info!(
+                    "{}: {}",
+                    self.plugin_name,
+                    std::str::from_utf8(&self.buffer.make_contiguous().split_last().unwrap().1)
+                        .unwrap()
+                );
+                self.buffer.clear();
+            }
+        }
+        Ok(buf.len())
+    }
```

Comment:
I think we need to add code here to ensure that we detect the 'log boundaries' correctly. For example, the plugin might call `write` on `stderr` in the middle of a log message. In such case we shouldn't call `info!` but buffer the bytes and wait until we the buffer on our side gets a `\n`. Also, I am not sure if current way of calling `from_utf` is safe. The `buf` might end on incorrect utf8 symbol boundary. Lastly, we probably should add some unit tests for this to make sure we're handling all those cases described above.

Reply:
We don't do it. I added comment in `DecoratingPipe::flush`.